docs/Configure.md
Kafka-php Configuration
==================

| Property	| C/P	| Range	| Default | Desc |
| --  | -- | -- | -- | -- |
| brokerVersion | C/P | 0.8.0 | 0.10.1.0 | User supplied broker version |
| clientId | C/P |  | kafka-php | This is a user supplied identifier for the client application |
| messageMaxBytes | C/P | 1000 .. 1000000000 | 1000000 | Maximum transmit message size. |
| metadataBrokerList | C/P | | | Kafka Broker server list |
| metadataMaxAgeMs | C/P | 1 .. 86400000 | -1 | Metadata cache max age. Defaults to metadata.refresh.interval.ms * 3 |
| metadataRefreshIntervalMs | C/P | 10 .. 3600000  | 300000 | Topic metadata refresh interval in milliseconds. The metadata is automatically refreshed on error and connect. Use -1 to disable the intervalled refresh.  |
| metadataRequestTimeoutMs | C/P | 10 .. 900000 | 60000 | Non-topic request timeout in milliseconds. This is for metadata requests, etc. |
| sslEnable | C/P | true/false | false | Whether enable ssl connect or not |
| sslEnableAuthentication | C/P | true/false | true | Enable authentication via SSL |
| sslCafile | C/P |  |  | Location of Certificate Authority file on local filesystem which should be used with the verify_peer context option to authenticate the identity of the remote peer.|
| sslLocalCert | C/P | File path |  | Path to local certificate file on filesystem. |
| sslLocalPk | C/P | File path |  | Path to local private key file on filesystem in case of separate files for certificate (local_cert) and private key. |
| sslPassphrase | C/P |  |  | Passphrase with which your local_cert file was encoded. |
| sslPeerName | C/P |  |  | Peer name to be used. If this value is not set, then the name is guessed based on the hostname used when opening the stream. |
| sslVerifyPeer | C/P | true/false | false | Require verification of SSL certificate used. |
| offsetReset | C | latest,earliest | latest | Action to take when there is no initial offset in offset store or the desired offset is out of range |
| groupId | C |  | |  Client group id string. All clients sharing the same group.id belong to the same group. |
| maxBytes | C |  | 65536 | Maximum bytes to fetch. |
| maxWaitTime | C |  | 100 | Maximum time in ms to wait for the response |
| sessionTimeout | C | 1 .. 3600000 | 30000 | Client group session and failure detection timeout.  |
| rebalanceTimeout | C | 1 .. 3600000 | 30000 | rebalance join wait timeout |
| topics | C | | |  Want consumer topics |
| isAsyn | P | true, false | false | Whether to use asynchronous production messages |
| produceInterval | P | 1 .. 900000 | 100 | The time interval at which requests for production messages are executed when the message is produced asynchronously |
| requestTimeout | P | 1 .. 900000 | 6000 |  The total timeout of the production message, which must be greater than the timeout config parameter |
| requiredAck | P | -1 .. 1000 | 1 | This field indicates how many acknowledgements the leader broker must receive from ISR brokers before responding to the request: 0=Broker does not send any response/ack to client, 1=Only the leader broker will need to ack the message, -1 or all=broker will block until message is committed by all in sync replicas (ISRs) or broker\'s in.sync.replicas setting before sending response.  |
| timeout | P | 1 .. 900000 | 5000 | Producer request timeout |

#### Note

All of the above parameters are set by setXxx, for example, to set the `clientId` parameter

```php
<?php

$config = \Kafka\ConsumerConfig::getInstance();
$config->setClientId('test');
```

Whether it is a consumer module or a production module, if the parameter settings do not match the rules will throw `\Kafka\Exception\Config` exception


docs/ch/Configure.md
Kafka-php 配置参数
==================

| Property	| C/P	| Range	| Default | Desc |
| ----  | ---- | ---- | ---- | ---- | 
| clientId | C/P | | kafka-php | 客户端标识 | 
| brokerVersion | C/P | 大于 0.8.0 | 0.10.1.0 | 为了计算 Kafka 请求的协议版本 |
| metadataBrokerList | C/P | | | 指定 Kafka Broker 列表，多个用逗号分隔 |
| messageMaxBytes | C/P | 1000 .. 1000000000 | 1000000 | 消息最大长度 |
| metadataRequestTimeoutMs | C/P | 10 .. 900000 | 60000 | 获取 meta 信息超时时间 |
| metadataRefreshIntervalMs | C/P | 10 .. 3600000  | 300000 | 获取同步 meta 信息的时间间隔 |
| metadataMaxAgeMs | C/P | 1 .. 86400000 | -1 | meta 信息有效期
| sslEnable | C/P | true/false | false | 是否开启 Ssl 连接 |
| sslEnableAuthentication | C/P | true/false | true | 通过SSL启用身份验证 |
| sslLocalCert | C/P | File path |  | 本地证书路径 |
| sslLocalPk | C/P | File path |  | 如果使用独立的文件来存储证书（local_cert）和私钥， 那么使用此选项来指明私钥文件的路径。|
| sslVerifyPeer | C/P | true/false | false | 是否需要验证 SSL 证书。|
| sslPassphrase | C/P |  |  | local_cert 文件的密码。|
| sslCafile | C/P |  |  | 当设置 sslVerifyPeer 为 true 时， 用来验证远端证书所用到的 CA 证书。 本选项值为 CA 证书在本地文件系统的全路径及文件名。|
| sslPeerName | C/P |  |  | 要连接的服务器名称。如果未设置，那么服务器名称将根据打开 SSL 流的主机名称猜测得出。|
| groupId | C |  | |  消费模块的分组 ID |
| sessionTimeout | C | 1 .. 3600000 | 30000 | 分组中消费者的有效时间 |
| rebalanceTimeout | C | 1 .. 3600000 | 30000 | 分组 rebalance 等待 join 时间 |
| topics | C | | |  将要消费的 kafka topic 名称 | 
| offsetReset | C | latest,earliest | latest | 如果消费 offset 失效的时候重置 offset 的策略 |
| maxBytes | C |  | 65536 | 单次 FETCH 请求对于单个分区请求的最大字节数 |
| maxWaitTime | C |  | 100 | 等待服务端响应 FETCH 请求的最大时间 |
| requiredAck | P | -1 .. 1000 | 1 | 生产消息确认策略 |
| timeout | P | 1 .. 900000 | 5000 | 生产消息请求超时时间 |
| isAsyn | P | true, false | false | 是否采用异步生产消息 |
| requestTimeout | P | 1 .. 900000 | 6000 |  消费消息整体超时时间, 该值一定要大于 timeout 参数 |
| produceInterval | P | 1 .. 900000 | 100 | 在异步生产消息时执行生产消息的请求的时间间隔 |

#### 注意

如上所有的参数均通过 setXxx 方式设置，例如要设置 `clientId` 参数

```php
<?php

$config = \Kafka\ConsumerConfig::getInstance();
$config->setClientId('test');
```

无论是消费模块还是生产模块，如果参数设置不符合规则时都会抛 `\Kafka\Exception\Config` 异常



src/Config.php
<?php
declare(strict_types=1);

namespace Kafka;

use function array_filter;
use function array_shift;
use function count;
use function explode;
use function in_array;
use function is_file;
use function lcfirst;
use function preg_match;
use function strpos;
use function substr;
use function trim;
use function version_compare;

/**
 * @method string getClientId()
 * @method string getBrokerVersion()
 * @method string getMetadataBrokerList()
 * @method int getMessageMaxBytes()
 * @method int getMetadataRequestTimeoutMs()
 * @method int getMetadataRefreshIntervalMs()
 * @method int getMetadataMaxAgeMs()
 * @method string getSecurityProtocol()
 * @method bool getSslEnable()
 * @method void setSslEnable(bool $sslEnable)
 * @method string getSslLocalCert()
 * @method string getSslLocalPk()
 * @method bool getSslVerifyPeer()
 * @method void setSslVerifyPeer(bool $sslVerifyPeer)
 * @method string getSslPassphrase()
 * @method void setSslPassphrase(string $sslPassphrase)
 * @method string getSslCafile()
 * @method string getSslPeerName()
 * @method void setSslPeerName(string $sslPeerName)
 * @method string getSaslMechanism()
 * @method string getSaslUsername()
 * @method string getSaslPassword()
 * @method string getSaslKeytab()
 * @method string getSaslPrincipal()
 */
abstract class Config
{
    public const SECURITY_PROTOCOL_PLAINTEXT      = 'PLAINTEXT';
    public const SECURITY_PROTOCOL_SSL            = 'SSL';
    public const SECURITY_PROTOCOL_SASL_PLAINTEXT = 'SASL_PLAINTEXT';
    public const SECURITY_PROTOCOL_SASL_SSL       = 'SASL_SSL';

    public const SASL_MECHANISMS_PLAIN         = 'PLAIN';
    public const SASL_MECHANISMS_GSSAPI        = 'GSSAPI';
    public const SASL_MECHANISMS_SCRAM_SHA_256 = 'SCRAM_SHA_256';
    public const SASL_MECHANISMS_SCRAM_SHA_512 = 'SCRAM_SHA_512';

    private const ALLOW_SECURITY_PROTOCOLS = [
        self::SECURITY_PROTOCOL_PLAINTEXT,
        self::SECURITY_PROTOCOL_SSL,
        self::SECURITY_PROTOCOL_SASL_PLAINTEXT,
        self::SECURITY_PROTOCOL_SASL_SSL,
    ];

    private const ALLOW_MECHANISMS = [
        self::SASL_MECHANISMS_PLAIN,
        self::SASL_MECHANISMS_GSSAPI,
        self::SASL_MECHANISMS_SCRAM_SHA_256,
        self::SASL_MECHANISMS_SCRAM_SHA_512,
    ];

    /**
     * @var mixed[]
     */
    protected static $options = [];

    /**
     * @var mixed[]
     */
    private static $defaults = [
        'clientId'                  => 'kafka-php',
        'brokerVersion'             => '0.10.1.0',
        'metadataBrokerList'        => '',
        'messageMaxBytes'           => 1000000,
        'metadataRequestTimeoutMs'  => 60000,
        'metadataRefreshIntervalMs' => 300000,
        'metadataMaxAgeMs'          => -1,
        'securityProtocol'          => self::SECURITY_PROTOCOL_PLAINTEXT,
        'sslEnable'                 => false, // this config item will override, don't config it.
        'sslEnableAuthentication'   => true,
        'sslLocalCert'              => false,
        'sslLocalPk'                => false,
        'sslVerifyPeer'             => false,
        'sslPassphrase'             => false,
        'sslCafile'                 => '',
        'sslPeerName'               => '',
        'saslMechanism'             => self::SASL_MECHANISMS_PLAIN,
        'saslUsername'              => '',
        'saslPassword'              => '',
        'saslKeytab'                => '',
        'saslPrincipal'             => '',
    ];

    /**
     * @param mixed[] $args
     *
     * @return bool|mixed
     */
    public function __call(string $name, array $args)
    {
        $isGetter = strpos($name, 'get') === 0 || strpos($name, 'iet') === 0;
        $isSetter = strpos($name, 'set') === 0;

        if (! $isGetter && ! $isSetter) {
            return false;
        }

        $option = lcfirst(substr($name, 3));

        if ($isGetter) {
            if (isset(self::$options[$option])) {
                return self::$options[$option];
            }

            if (isset(self::$defaults[$option])) {
                return self::$defaults[$option];
            }

            if (isset(static::$defaults[$option])) {
                return static::$defaults[$option];
            }

            return false;
        }

        if (count($args) !== 1) {
            return false;
        }

        static::$options[$option] = array_shift($args);

        // check todo
        return true;
    }

    /**
     * @throws Exception\Config
     */
    public function setClientId(string $val): void
    {
        $client = trim($val);

        if ($client === '') {
            throw new Exception\Config('Set clientId value is invalid, must is not empty string.');
        }

        static::$options['clientId'] = $client;
    }

    /**
     * @throws Exception\Config
     */
    public function setBrokerVersion(string $version): void
    {
        $version = trim($version);

        if ($version === '' || version_compare($version, '0.8.0', '<')) {
            throw new Exception\Config('Set broker version value is invalid, must is not empty string and gt 0.8.0.');
        }

        static::$options['brokerVersion'] = $version;
    }

    /**
     * @throws Exception\Config
     */
    public function setMetadataBrokerList(string $brokerList): void
    {
        $brokerList = trim($brokerList);

        $brokers = array_filter(
            explode(',', $brokerList),
            function (string $broker): bool {
                return preg_match('/^(.*:[\d]+)$/', $broker) === 1;
            }
        );

        if (empty($brokers)) {
            throw new Exception\Config(
                'Broker list must be a comma-separated list of brokers (format: "host:port"), with at least one broker'
            );
        }

        static::$options['metadataBrokerList'] = $brokerList;
    }

    public function clear(): void
    {
        static::$options = [];
    }

    /**
     * @throws Exception\Config
     */
    public function setMessageMaxBytes(int $messageMaxBytes): void
    {
        if ($messageMaxBytes < 1000 || $messageMaxBytes > 1000000000) {
            throw new Exception\Config('Set message max bytes value is invalid, must set it 1000 .. 1000000000');
        }
        static::$options['messageMaxBytes'] = $messageMaxBytes;
    }

    /**
     * @throws Exception\Config
     */
    public function setMetadataRequestTimeoutMs(int $metadataRequestTimeoutMs): void
    {
        if ($metadataRequestTimeoutMs < 10 || $metadataRequestTimeoutMs > 900000) {
            throw new Exception\Config('Set metadata request timeout value is invalid, must set it 10 .. 900000');
        }
        static::$options['metadataRequestTimeoutMs'] = $metadataRequestTimeoutMs;
    }

    /**
     * @throws Exception\Config
     */
    public function setMetadataRefreshIntervalMs(int $metadataRefreshIntervalMs): void
    {
        if ($metadataRefreshIntervalMs < 10 || $metadataRefreshIntervalMs > 3600000) {
            throw new Exception\Config('Set metadata refresh interval value is invalid, must set it 10 .. 3600000');
        }
        static::$options['metadataRefreshIntervalMs'] = $metadataRefreshIntervalMs;
    }

    /**
     * @throws Exception\Config
     */
    public function setMetadataMaxAgeMs(int $metadataMaxAgeMs): void
    {
        if ($metadataMaxAgeMs < 1 || $metadataMaxAgeMs > 86400000) {
            throw new Exception\Config('Set metadata max age value is invalid, must set it 1 .. 86400000');
        }
        static::$options['metadataMaxAgeMs'] = $metadataMaxAgeMs;
    }

    /**
     * @throws Exception\Config
     */
    public function setSslLocalCert(string $localCert): void
    {
        if (! is_file($localCert)) {
            throw new Exception\Config('Set ssl local cert file is invalid');
        }

        static::$options['sslLocalCert'] = $localCert;
    }

    /**
     * @throws Exception\Config
     */
    public function setSslLocalPk(string $localPk): void
    {
        if (! is_file($localPk)) {
            throw new Exception\Config('Set ssl local private key file is invalid');
        }

        static::$options['sslLocalPk'] = $localPk;
    }

    /**
     * @throws Exception\Config
     */
    public function setSslCafile(string $cafile): void
    {
        if (! is_file($cafile)) {
            throw new Exception\Config('Set ssl ca file is invalid');
        }

        static::$options['sslCafile'] = $cafile;
    }

    /**
     * @throws Exception\Config
     */
    public function setSaslKeytab(string $keytab): void
    {
        if (! is_file($keytab)) {
            throw new Exception\Config('Set sasl gssapi keytab file is invalid');
        }

        static::$options['saslKeytab'] = $keytab;
    }

    /**
     * @throws Exception\Config
     */
    public function setSecurityProtocol(string $protocol): void
    {
        if (! in_array($protocol, self::ALLOW_SECURITY_PROTOCOLS, true)) {
            throw new Exception\Config('Invalid security protocol given.');
        }

        static::$options['securityProtocol'] = $protocol;
    }

    /**
     * @throws Exception\Config
     */
    public function setSaslMechanism(string $mechanism): void
    {
        if (! in_array($mechanism, self::ALLOW_MECHANISMS, true)) {
            throw new Exception\Config('Invalid security sasl mechanism given.');
        }

        static::$options['saslMechanism'] = $mechanism;
    }
}
